src/Orleans/Configuration/GlobalConfiguration.cs

```diff
@@ -217,21 +219,35 @@ public bool PrimaryNodeIsRequired
 
         #region MultiClusterNetwork
 
+        private string clusterId;
+
         /// <summary>
         /// Whether this cluster is configured to be part of a multicluster network
         /// </summary>
         public bool HasMultiClusterNetwork
         {
             get
             {
-                return !(string.IsNullOrEmpty(ClusterId));
+                return !(string.IsNullOrEmpty(this.clusterId));
             }
         }
 
         /// <summary>
         /// Cluster id (one per deployment, unique across all the deployments/clusters)
         /// </summary>
-        public string ClusterId { get; set; }
+        public string ClusterId
+        {
+            get
+            {
+                var configuredId = this.HasMultiClusterNetwork ? this.clusterId : this.DeploymentId;
+                return string.IsNullOrEmpty(configuredId) ? DefaultClusterId : configuredId;
+            }
+
+            set
+            {
+                this.clusterId = value;
+            }
+        }
```

Comment:
I guess it's fine, when I first added the comment I thought this was just new logic, didn't know it was moved from somewhere else. I agree we should eventually move it somewhere else, since the config is no longer some config holder, but it's fine for now, since we'll revamp the whole config object anyway.